### PR TITLE
[DRAFT] <amp-embed type="insticator">

### PR DESCRIPTION
**Pull request by vladdeev**
This is a draft implementation of `<amp-embed type=“insticator”>`. 
The intended functionality is described in #23257. 
We need technical advice before going further.

**Implementation descriptionю**
The `<amp-embed>` creates an iframe with a boilerplate document inside.
We add three children to the `div#c` in that document in the following order:
*amp-ad which holds the upper ad
*iframe which holds our embed
*amp-ad which holds the lower ad

cc @lannka

### CODE REVIEW

File: ads/insticator.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+//Don't forget to preconnect and prefetch as it's described in the docs
+//https://github.com/ampproject/amphtml/blob/master/ads/README.md
+const url = {
+  ads: 'https://drhn9v8cwg89y.cloudfront.net',
+  embed: 'https://d3lcz8vpax4lo2.cloudfront.net',
+  lib: 'https://testthisshit.online/amp-embed-lib/insticator.js'
+};
+
+// ------- HELPER FUNCTIONS ------- //
+function createInitialMarkup() {
```
Comment (lannka) [inline ads/insticator.js]: for flexibility in a long run, I suggest you only leave config related logic in this file, and move logic to `/amp-embed-lib/insticator.js` as much as possible.

**Comment by ortemcheg** [comment]
Currently the ads won’t render because the boilerplate HTML they get appended to is not a valid AMP. 
One way to make it work is to use iframes with valid AMP in place of `<amp-ad>`s. If we take that path we’ll end up with three iframes placed inside of an iframe created by `<amp-embed>`. We think such an implementation would be ugly, slow and inefficient, so we need to implement it differently. So now we badly need suggestions from the AMP team to help us with implementation. @lannka, could you help?

Because the only alternative way of implementing this functionality that we see now is to consider once again `<amp-insticator>`, which doesn’t have any nested iframes and uses native `<amp-ads>` organically.

**Comment by triso07** [comment]
@lannka Any help you can provide on the above would be great. We're starting to get under the gun in terms of our build effort here. Need to start making consistent forward progress now.

**Comment by triso07** [comment]
@lannka Maybe we can just hop on a call if you're open to that? That may speed up the process and make this easier for all of us.

**Comment by triso07** [comment]
@lannka Please advise on the best course of action here

**Comment by triso07** [comment]
@lannka Can we just get on a call to hash this out? We can probably clear this up very quickly and move forward. This starting and stopping of development is hurting our progress and we have very real business needs we're trying to meet here.

**Comment by triso07** [comment]
This back and forth messaging feels a bit slow and unproductive. A call would help expedite this process. We have technical concerns with `<amp-embed>`. Unless we can resolve those `<amp-insticator>` seems a much more viable implementation.

**Comment by lannka** [comment]
sorry for the delay. VC sounds good. Are you in GMT+3? I'm in PST. Try to find a good time for both.

**Comment by lannka** [comment]
BTW, inside the iframe, why do you still need `amp-ad`? You can have your own ads tag to retrieve ads. For sake of performance, you might consider server-to-server fetch to reduce client side redirect.

**Comment by triso07** [comment]
Thanks for the response. We're EST. If you have time tomorrow that would be great. How's 4:30pm EST?

**Comment by lannka** [comment]
chatted over VC, meeting notes:

- `<amp-embed>` uses 3rd party iframe, inside the iframe Instricator can do things just like non-AMP pages, with some limitation as noted below
- To prevent page jump, like any other AMP extension, `<amp-embed>` is not allowed to resize itself freely. Instead, 3rd party code should call the context API to do resize. The context API also provides a bunch of commonly used API for 3rd party code to retrieve relevant info about the apge.
- The code snippet that Instricator provides to their customers should look like: `<amp-embed type=instricator width=300 height=750>`. That said, a predefined size need to be specified.
- Elements within current viewport are generally not allowed for resize. To increase success rate of a resize, do it as early as possible before user scroll to it.
- On a resize failure, Instricator can consider to use a overflow button to do a user initiated resize.

**Comment by triso07** [comment]
@lannka Can you please review our component when you have a moment? We feel it's in a good place.

**Comment by triso07** [comment]
@lannka Can you please review and let me know if we're good?

File: ads/insticator.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+};
+
+
+
+/*
+ * @param {!Window} global
+ * @param {!Object} data
+*/
+// ------- AMP EMBED - INSTICATOR EXPORT ------- //
+export function insticator(global, data) {
```
Comment (lannka) [inline ads/insticator.js]: code style: public function at top before all private functions

File: ads/insticator.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+// Don't forget to preconnect and prefetch as it's described in the docs
+// https://github.com/ampproject/amphtml/blob/master/ads/README.md
+const url = {
+  content: '//d3lcz8vpax4lo2.cloudfront.net',
+  lib: 'https://testthisshit.online/amp-embed-lib/insticator.js'
+};
+
+
+// ------- HELPER FUNCTIONS ------- //
+const createInitialMarkup = (embedId) => {
```
Comment (lannka) [inline ads/insticator.js]: code style: please use `function createInitialMarkup(embedId) {`

File: ads/insticator.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  return `
+    <div id="insticator-container">
+      <div id="div-insticator-ad-1"></div>
+      <div id="insticator-embed" embed-id="${embedId}"></div>
+      <div id="div-insticator-ad-2"></div>
+    </div>
+  `;
+};
+
+const createTemplate = (domString) => {
```
Comment (lannka) [inline ads/insticator.js]: can this be merged into `createInitialMarkup`

File: ads/insticator.js (hunk 2 above)
Comment (triso07) [reply ads/insticator.js]: Updated

File: ads/insticator.js (hunk 3 above)
Comment (triso07) [reply ads/insticator.js]: Updated

File: ads/insticator.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  const template = document.createElement('template');
+  template.innerHTML = domString;
+  return template.content;
+};
+
+const appendElement = (location, el) => {
```
[the remark this answers is not in the document]
Comment (triso07) [reply ads/insticator.js]: Removed this function

File: ads/insticator.js (hunk 4 above)
Comment (triso07) [reply ads/insticator.js]: Merged code

File: ads/insticator.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  // API available to embed
+  // https://github.com/ampproject/amphtml/blob/master/ads/README.md#available-information-to-the-ad
+  // console.log(window.context)
+
+  // create virtual script tag elements
+  const headerCode = createElement(componentContainer.ownerDocument, 'script', { type: 'text/javascript' });
+  const bodyCode = createElement(componentContainer.ownerDocument, 'script', { type: 'text/javascript' });
+
+  // populate script tag elements with standard insticator embed code (slightly modified header code to account for AMP specific constraints)
+  headerCode.appendChild(document.createTextNode(`
```
[the remark this answers is not in the document]
Comment (triso07) [reply ads/insticator.js]: Adjusted to execute instead of append script elements

**Comment by triso07** [comment]
@lannka Changes made. Please have a look when you have a moment.

File: ads/insticator.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  b.src = u;
+  b.async = !0;
+  const d = c.getElementsByTagName(s)[0];
+  d.parentNode.insertBefore(b, d);
+
+  // execute functions of insticator object on the window (load ads and embed)
+  Insticator.ad.loadAd("div-insticator-ad-1");
+  Insticator.ad.loadAd("div-insticator-ad-2");
+  Insticator.load("em", {id : embedId});
+}
```
Comment (lannka) [inline ads/insticator.js]: new line at the end of the file

File: ads/_config.js [hunk 8]
```diff
@@ -534,6 +534,13 @@ const adConfig = jsonConfiguration({
     renderStartImplemented: true,
   },
 
+  'insticator': {
+    prefetch: 'https://testthisshit.online/amp-embed-lib/insticator.js',
+    preconnect: [
+      'https://d3lcz8vpax4lo2.cloudfront.net'
+    ]
+   },
```
Comment (lannka) [inline ads/_config.js]: would you please also implement renderStart API?

in your remote loaded `insticator.js`, it should call `context.renderStart` once content is ready to show. This basically gives AMP runtime signal when to hide loading indicator and show your content.

File: ads/insticator.js (hunk 7 above)
Comment (triso07) [reply ads/insticator.js]: New line added at the end of the document

File: ads/_config.js (hunk 8 above)
Comment (triso07) [reply ads/_config.js]: renderStart implemented

**Comment by triso07** [comment]
@lannka Changes complete. Let me know if this is good to go.

**Comment by triso07** [comment]
@lannka Hey there. It's been about a week, just want to check-in and see if we're good to go with these changes.

**Comment by calebcordry** [comment]
@triso07 Attached screenshot of the example creative, is this what we would expect to see?
image

**Comment by calebcordry** [comment]
@triso07 I also sometimes see this, seems you would like a larger size? Did you want to add a overflow button as @lannka mentioned earlier?
image

File: ads/_config.js (hunk 8 above)
Comment (calebcordry) [reply ads/_config.js]: 👍verified amp runtime is getting this signal.

File: ads/insticator.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
+// !!!!!!!!!! ---------- AMP EMBED - INSTICATOR ---------- !!!!!!!!!! //
```
Comment (calebcordry) [inline ads/insticator.js]: please add the amp license header to this file.

File: ads/insticator.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import { validateData, loadScript } from '../3p/3p';
+
+
+/*
+ * @param {!Window} global
+ * @param {!Object} data
+*/
+// ------- AMP EMBED - INSTICATOR EXPORT ------- //
```
Comment (calebcordry) [inline ads/insticator.js]: Our linter doesn't like this comment, could you move it to the JSDoc block?

File: ads/insticator.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+// ------- HELPER VARIABLES ------- //
+// reusable URL references to ads, embed and the library
+// Don't forget to preconnect and prefetch as it's described in the docs
+// https://github.com/ampproject/amphtml/blob/master/ads/README.md
+const url = { content: '//d3lcz8vpax4lo2.cloudfront.net' };
+
+
+// ------- HELPER FUNCTIONS ------- //
+function createTemplate(embedId) {
```
Comment (calebcordry) [inline ads/insticator.js]: nit: jsdoc

File: ads/insticator.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    <div id="insticator-container">
+      <div id="div-insticator-ad-1"></div>
+      <div id="insticator-embed" embed-id="${embedId}"></div>
+      <div id="div-insticator-ad-2"></div>
+    </div>
+  `;
+  return template.content;
+}
+
+function createAdsAndEmbed(siteId, embedId) {
```
Comment (calebcordry) [inline ads/insticator.js]: nit: jsdoc

**Comment by triso07** [comment]
Thanks for the response. Will make adjustments and get back to you.

File: ads/insticator.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  const a = window;
+  const c = document;
+  const s = 'script';
+  const u = `${url.content}/ads-code/${siteId}.js`; // vars from preconnect urls and data attributes on amp-embed tag
+
+  // create insticator object on the window
+  'Insticator' in a || (a.Insticator = {
+      ad: {
+          loadAd: function(b) {
+              Insticator.ad.q.push(b)
```
Comment (calebcordry) [inline ads/insticator.js]: TIL this is valid JS, however our linter doesn't like it. How about `this.ad.q.push...`

File: ads/insticator.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+          },
+          q: []
+      },
+      helper: {},
+      embed: {},
+      version: "4.0",
+      q: [],
+      amp: null, // this will get set to window.context which is the AMP API so we can access from our ads code
+      load: function(t, o) {
+          Insticator.amp = window.context; // set the Insticator object property amp to window.context which is the AMP API so we can access from our ads code
```
Comment (calebcordry) [inline ads/insticator.js]: Same here and next line. Linter will complain about any use of Insticator within this obj.

**Comment by calebcordry** [comment]
Also if you are able to run `gulp lint` locally, it should catch a few of the nits I mentioned.

**Comment by triso07** [comment]
@calebcordry Fixed all linter callouts. Should be good. Please have a look when you have a moment.

File: ads/insticator.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+  const d = c.getElementsByTagName(s)[0];
+  d.parentNode.insertBefore(b, d);
+
+  // execute functions of insticator object on the window (load ads and embed)
+  a.Insticator.ad.loadAd('div-insticator-ad-1');
+  a.Insticator.ad.loadAd('div-insticator-ad-2');
+  a.Insticator.load('em', {id: embedId});
+
+  // now tell AMP runtime to start rendering ads
+  window.context.renderStart();
```
Comment (calebcordry) [inline ads/insticator.js]: I see that you moved this into this file, I think you were calling it from the external JS before, did you remove it from the external JS? Just want to make sure you are now not calling it twice.

File: ads/insticator.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+const url = {content: '//d3lcz8vpax4lo2.cloudfront.net'};
+
+/**
+ * Create HTML template to be populated later
+ * @param {string} embedId The Unique Identifier of this particular Embed
+ * @return {element} HTML template
+ */
+function createTemplate(embedId) {
+  const template = document.createElement('template');
+  template.innerHTML = `
```
Comment (calebcordry) [inline ads/insticator.js]: you will need to mark this with `template./*OK*/.innnerHTML`

**Comment by calebcordry** [comment]
Thanks for working on this. I am now getting a creative similar to https://github.com/ampproject/amphtml/pull/23928#issuecomment-536662558 every time. Is this what you all want? I just want to make sure the sizing is what you expect?

File: ads/insticator.js [hunk 17]
```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
```
Comment (calebcordry) [inline ads/insticator.js]: nit: 2019

File: ads/insticator.js (hunk 15 above)
Comment (triso07) [reply ads/insticator.js]: I have `renderStartImplemented: true,` in _config.js. Then I'm calling it in insticator.js. Is that the correct implementation for this? I don't think we're calling this from any external JS file.

File: ads/insticator.js (hunk 16 above)
Comment (triso07) [reply ads/insticator.js]: When I add this I get an error in the console ...

`SyntaxError: /Users/troyriso/Desktop/workspace/amphtml/ads/insticator.js: Unexpected token (50:17)

  48 | function createTemplate(embedId) {
  49 |   const template = document.createElement('template');
> 50 |   template./*OK*/.innnerHTML = `
     |                  ^
  51 |     <div id="insticator-container">
  52 |       <div id="div-insticator-ad-1"></div>
  53 |       <div id="insticator-embed" embed-id="${embedId}"></div> while parsing file: /Users/troyriso/Desktop/workspace/amphtml/ads/insticator.js
[14:12:15] Error while compiling integration.js`

File: ads/insticator.js [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+  validateData(data, ['siteId', 'embedId']);
+
+  // create insticator markup
+  global.document.getElementById('c').appendChild(createTemplate(data.embedId));
+
+  // create ads and embed
+  createAdsAndEmbed(data.siteId, data.embedId);
+
+  // envoke AMP library
+  loadScript(global);
```
Comment (calebcordry) [inline ads/insticator.js]: I think this is not doing anything?

File: ads/insticator.js (hunk 15 above)
Comment (calebcordry) [reply ads/insticator.js]: Ok this was probably my misunderstanding. This is fine.

File: ads/insticator.js (hunk 16 above)
Comment (calebcordry) [reply ads/insticator.js]: ah sorry typo. There shouldn't be a second `.` 
`template./*OK*/innnerHTML`

File: ads/insticator.js (hunk 17 above)
Comment (triso07) [reply ads/insticator.js]: Updated to 2019

File: ads/insticator.js (hunk 18 above)
Comment (triso07) [reply ads/insticator.js]: Removed. Thought this was necessary from the AMP side.

File: ads/insticator.js (hunk 16 above)
Comment (triso07) [reply ads/insticator.js]: No worries. This one doesn't throw a console error, but now the embed doesn't load.

**Comment by triso07** [comment]
Thanks for the quick responses by the way

File: ads/insticator.js (hunk 16 above)
Comment (triso07) [reply ads/insticator.js]: Hmm, got it to work. Should be updated now.

File: ads/insticator.js [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+const url = {content: '//d3lcz8vpax4lo2.cloudfront.net'};
+
+/**
+ * Create HTML template to be populated later
+ * @param {string} embedId The Unique Identifier of this particular Embed
+ * @return {element} HTML template
+ */
+function createTemplate(embedId) {
+  const template = document.createElement('template');
+  template./*OK*/innerHTML = `
```
Comment (calebcordry) [inline ads/insticator.js]: ugh looks like you need a space here for linter `template./*OK/ innerHTML`

Comment (triso07) [reply ads/insticator.js]: Adding a space breaks the code and throws a console error

Comment (triso07) [reply ads/insticator.js]: @calebcordry Can I maybe just remove the /*OK*/ stuff? I'm unfamiliar with that. What does it do?

Comment (calebcordry) [reply ads/insticator.js]: We forbid the use of `innerHTML` in our code base unless reviewed. The `OK` lets us know someone has approved it. I just cloned your PR and made the change, and it seems to be fine? 
image

Comment (triso07) [reply ads/insticator.js]: Seems to be working for me now as well. Strange. 

This has now been updated.

**Comment by triso07** [comment]
@calebcordry After that innerHTML OK addition are we good to go?

**Comment by calebcordry** [comment]
Great you got the innerHTML figured out. Looks like you are now getting a few type errors (available by running `gulp check-types`). I will make notes in the code as well. 

Also it seems like there are now a few commits with authors that are not covered by the Google CLA, this will have to be resolved before we can merge.
Screen Shot 2019-10-04 at 9 37 40 AM

File: ads/insticator.js [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+  global.document.getElementById('c').appendChild(createTemplate(data.embedId));
+
+  // create ads and embed
+  createAdsAndEmbed(data.siteId, data.embedId);
+}
+
+/**
+ * Create HTML template to be populated later
+ * @param {string} embedId The Unique Identifier of this particular Embed
+ * @return {element} HTML template
```
Comment (calebcordry) [inline ads/insticator.js]: nit: `{Element}`

File: ads/insticator.js [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function insticator(global, data) {
+  // validate passed data attributes
+  validateData(data, ['siteId', 'embedId']);
+
+  // create insticator markup
+  global.document.getElementById('c').appendChild(createTemplate(data.embedId));
```
Comment (calebcordry) [inline ads/insticator.js]: nit: these data properties will need to be accessed using bracket notation. `data['embedId']`

File: ads/insticator.js [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+  // create ads and embed
+  createAdsAndEmbed(data.siteId, data.embedId);
```
Comment (calebcordry) [inline ads/insticator.js]: `data['siteId']` `data['embedId']`

File: ads/insticator.js (hunk 20 above)
Comment (triso07) [reply ads/insticator.js]: Fixed

File: ads/insticator.js (hunk 21 above)
Comment (triso07) [reply ads/insticator.js]: Fixed

File: ads/insticator.js (hunk 22 above)
Comment (triso07) [reply ads/insticator.js]: Fixed

**Comment by vladdeev** [comment]
> Also it seems like there are now a few commits with authors that are not covered by the Google CLA, this will have to be resolved before we can merge.
> Screen Shot 2019-10-04 at 9 37 40 AM

hey @calebcordry I added these 2 addresses to the group. Do we need to do anything else to pass this check?

**Comment by calebcordry** [comment]
Thanks for looking into the CLA. I do see the new emails passing but it looks like a previously ok email is now failing. 
image

**Comment by triso07** [comment]
@googlebot I consent

File: ads/insticator.js [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function insticator(global, data) {
+  // validate passed data attributes
+  validateData(data, ['siteId', 'embedId']);
+
+  // create insticator markup
+  global.document.getElementById('c').appendChild(createTemplate(data['embedId']));
```
Comment (calebcordry) [inline ads/insticator.js]: nit: linter wants these on new lines. If you run `gulp lint` it will show these errors.

Comment (triso07) [reply ads/insticator.js]: Updated

**Comment by triso07** [comment]
@calebcordry Great! Thanks!

**Comment by rcebulko** [comment]
Oddly, requesting the set of reviews from the GitHub API for this PR yields the following series of states: 

```javascript
[ [ 'COMMENTED', 'lannka' ],
  [ 'DISMISSED', 'lannka' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'lannka' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'calebcordry' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'triso07' ],
  [ 'COMMENTED', 'calebcordry' ] ]
```

Investigating why GitHub thinks you haven't approved @calebcordry

**Comment by calebcordry** [comment]
@rcebulko Thanks for investigating. 

@triso07 You are welcome. Thanks for being patient on this contribution! FYI we are investigating what is going on with our owners bot, no further changes are needed from you. I will merge when it gets sorted.

**Comment by calebcordry** [comment]
@triso07 do you have an open PR?

**Comment by triso07** [comment]
@calebcordry We do now ;)

https://github.com/ampproject/amphtml/pull/25103

**Pull request merged**